### PR TITLE
added the BundleInit function in package and BundleInitSpec for testi…

### PR DESCRIPTION
**Pull request by poddar92**
…ng it with one basic test.(doesnt work right now)

### Contributor Checklist

- [ ] Did you add Scaladoc to every public function/method?
- [y ] Did you add at least one test demonstrating the PR?
- [y ] Did you delete any extraneous printlns/debugging code?
- [y ] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [ ] Did you state the API impact?
- [ ] Did you specify the code generation impact?
- [ ] Did you request a desired merge strategy?
- [ ] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

 

- new feature/API

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (Bug fix: `3.4.x`, [small] API extension: `3.5.x`, API modification or big change: `3.6.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

**Comment by poddar92** [comment]
@jackkoenig I have just created a draft PR; so that we can discuss the issue I was seeing. This is also mentioned in the issue comment. #2435

**Comment by jackkoenig** [comment]
@poddar92 I apologize for not responding to your question in that comment, I must've missed the notification about it. Pasting your question below so we can continue discussion here:

> @jackkoenig do we want to recursively go though the bundle and then initialize them? what if one of the elements is a Vec? 
Should we follow how Bundle.Lit does it?
Also; i see the below error with above sketch:
>
> ```
> [error] /Users/radhikapoddar/chisel3/core/src/main/scala/chisel3/package.scala:395:24: macro implementation not found: materialize
> [error] (the most common reason for that is that you cannot use macro implementations in the same compilation run that defines them)
> [error]         val init = Wire(bun)
> ```
>
> is it because we are trying to assign the bundle directly without defining its type in Wire first? is this related to sourceInfo?
 
This issue is a bit subtle, and for better understanding, I suggest you watch my video describing some of the details of how Chisel's internals work: [Chisel Breakdown 3](https://www.youtube.com/watch?v=4vgSrW7Cdkk).

Basically, yes this is related to `SourceInfo` and the macro that materializes source locators. Both `Wire` and `:=` require a source locator but since you are in Chisel's internals, you're not allowed to materialize source locators pointing to the internals. The solution is to have the function you are adding take a `SourceInfo` implicit argument. Then, both `Wire` and `:=` will use _that_ source locator which will be materialized in the caller's source code. I'll leave a suggestion on your code showing what I mean.

File: core/src/main/scala/chisel3/package.scala [hunk 1]
```diff
@@ -412,4 +412,19 @@ package object chisel3 {
     "this feature will not be supported as part of the migration to the MLIR-based FIRRTL Compiler (MFC). For more information about this migration, please see the Chisel ROADMAP.md."
 
   final val deprecatedPublicAPIMsg = "APIs in chisel3.internal are not intended to be public"
+  
+  object BundleInit {
+    implicit class AddInitConstruct[T <: Record](bun: T) {
+      def Init(fs: (T => (Data, Data))*): T = {
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/package.scala:418]:
```suggestion
      def Init(fs: (T => (Data, Data))*)(implicit sourceInfo: SourceInfo): T = {
```
As I alluded to in my comment, https://github.com/chipsalliance/chisel/pull/3060#issuecomment-1452793826, you need to thread a source locator through so that it can be materialized in the caller's code.
